enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,265 @@
+
+#### Operand cleanup status
+
+To provide visibility on the status of operand cleanup, OLM can list out the CRs that are still pending deletion in the CSV status e.g `status.cleanup.pendingDeletion`.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+status:
+  ...
+  cleanup:
+    pendingDeletion:
+    - resource: etcdclusters.etcd.database.coreos.com
+      kind: EtcdCluster
+      version: v1beta2
+      name: Foo
+      namespace: Bar
+    - kind: EtcdCluster
+      ...
+```
+
+This information can be useful when debugging a stalled operator uninstall process to see the CRs whose removal is blocking operand cleanup.
+
+**NOTE:** The number of CRs that OLM has to list in the status is unbounded and could potentially cause the CSV size to exceed the [max request size][max-request-size] configured in etcd. For this reason there would have to be an upper limit on the number of CRs displayed in `status.cleanup.pendingDeletion`.
+
+The CSV can also be updated with a status condition to indicate that cleanup is in progress when the CSV is blocked on deletion.
+
+```YAML
+status:
+  conditions:
+  - lastTransitionTime: "..."
+    lastUpdateTime: "..."
+    message: |
+      waiting for operator to finish cleanup for 4 CRs
+    phase: Cleanup
+    reason: CleanupWaiting
+```
+
+This CSV will transition to a new phase `Cleanup` with the reason `CleanupWaiting` when it is pending deletion with the finalizer `operatorframework.io/cleanup-apis` present.
+
+#### Handling CSV deletion due to replacement
+
+A CSV can also be [deleted by OLM][csv-deletion-on-upgrade] during it's lifecycle when OLM detects a newer CSV that replaces it as part of an upgrade. For an operator that has already opted-in to operand cleanup, this would result in OLM cleaning up all the CRs before removing the old CSV.
+
+To prevent the cleanup of CRs on an upgrade, OLM should ensure that a CSV which has `spec.cleanup.enabled: true` and is marked for replacement (`phase: replacing`), should first be cleared of the `operatorframework.io/cleanup-apis` finalizer and opted-out of cleanup before it is marked for deletion or deleted.
+
+#### CRD removal
+
+While not strictly necessary for the goal of cleaning up operator-created resources, the cleanup process can also include removing the CRDs after the CRs and operator have been removed. The CRD can't always be removed however.
+
+If there are two non-global operators installed of the same type (providing the same APIs), then removing the CRD during the cleanup of one operator would result in the deletion of CRs for the other operator as well.
+
+To avoid this case OLM would have to ensure that it only removes the CRD during cleanup if:
```

Comment:
I think there's also the case where the CRD being deleted is a `required` CRD for another CSV -- in that case removing the CRD would break an operator that has a dependency on that CRD, even if they are unrelated. We should probably avoid deleting the CRD in that case as well.